### PR TITLE
Better support for conditional expressions in JSpecify mode

### PR DESCRIPTION
**Pull request by msridhar**
We improve support for conditional expressions in JSpecify mode, by properly inferring types of conditional expressions where appropriate.  We also propagate types from the assignment context through (nested) conditional expressions to correctly infer types of generic method calls nested within such expressions.  See the new `ConditionalExprTests` for examples.

We extract out some common code for finding a type from the assignment context given a `TreePath` into a new method `getTargetTypeFromParentContext`, and now use it consistently to handle getting assignment contexts for conditional expressions, generic method calls, and diamond constructor calls.  As part of this extraction, we make more consistent our handling of `var`-declared locals (ensuring we don't use their `javac`-inferred type during our own inference), and we add tests for this.

Apologies for the size of this PR; it would have been a bit awkward to make it smaller.

## Release Notes

* **New Features**
  * Improved nullability and generic type inference for conditional (`?:`) expressions, including nested conditionals and scenarios influenced by dataflow/loops.
  * More accurate type inference when conditional results feed into assignments, method calls, and diamond constructor usage.
  * Enhanced handling of `var` cases where conditional expressions don’t supply enough target type information for generic/diamond inference.

* **Tests**
  * Added expanded JUnit coverage for conditional-expression inference/nullability, diamond and generic inference edge cases, and dataflow/loop interactions.

### CODE REVIEW

**Comment by msridhar** [comment]
This change is part of the following stack:

- #1608 ◀
    - #1627

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 1]
```diff
@@ -1816,42 +1810,261 @@ public void checkTypeParameterNullnessForConditionalExpression(
...
+      TreePath enclosingMethodOrLambda =
+          NullabilityUtil.findEnclosingMethodOrLambdaOrInitializer(parentPath);
+      if (enclosingMethodOrLambda != null
+          && enclosingMethodOrLambda.getLeaf() instanceof MethodTree enclosingMethod) {
+        Symbol.MethodSymbol methodSymbol = ASTHelpers.getSymbol(enclosingMethod);
+        if (methodSymbol != null) {
+          return new TargetTypeAndAssignmentKind(methodSymbol.getReturnType(), false);
+        }
+      }
+      return new TargetTypeAndAssignmentKind(null, false);
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java]: Maybe silly (and valid defensive programming), but isn't this the same as the full method's default case? I think removing this return preserves the same method logic. Unless I am missing something.

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 2]
```diff
@@ -1816,42 +1810,261 @@ public void checkTypeParameterNullnessForConditionalExpression(
...
+        return new TargetTypeAndAssignmentKind(null, false);
+      }
+      Type methodType = ASTHelpers.getType(parentInvocation.getMethodSelect());
+      if (methodType == null) {
+        return new TargetTypeAndAssignmentKind(null, false);
+      }
+      return new TargetTypeAndAssignmentKind(
+          getFormalParameterTypeForArgument(
+              parentInvocation, methodType.asMethodType(), expressionTree),
+          false);
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java]: Similarly, this could be...
```suggestion
      if (methodType != null) {
        return new TargetTypeAndAssignmentKind(
          getFormalParameterTypeForArgument(
              parentInvocation, methodType.asMethodType(), expressionTree),
          false);
      }
```

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 3]
```diff
@@ -1816,42 +1810,261 @@ public void checkTypeParameterNullnessForConditionalExpression(
...
+   * from the initializer.
+   */
+  private TargetTypeAndAssignmentKind getTargetTypeFromParentContext(
+      Tree tree, TreePath path, VisitorState state, boolean calledFromDataflow) {
+    Tree expressionTree = tree;
+    TreePath expressionPath = path;
+    TreePath parentPath = expressionPath.getParentPath();
+    if (parentPath == null) {
+      return new TargetTypeAndAssignmentKind(null, false);
+    }
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:1956]:
```suggestion
    }
    // 1. Climb up to corresponding "assignment-like" context:
```

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 4]
```diff
@@ -1816,42 +1810,261 @@ public void checkTypeParameterNullnessForConditionalExpression(
...
+      expressionPath = getOutermostConditionalExpressionPath(parentPath);
+      expressionTree = expressionPath.getLeaf();
+      parentPath = expressionPath.getParentPath();
+      parent = parentPath.getLeaf();
+      while (parent instanceof ParenthesizedTree) {
+        parentPath = parentPath.getParentPath();
+        parent = parentPath.getLeaf();
+      }
+    }
+    VisitorState parentState = state.withPath(parentPath);
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:1973]:
```suggestion
    VisitorState parentState = state.withPath(parentPath);
    // 2. We are at the right point in the tree now. Three cases:
    // 2a. `foo = [expr];` or `var foo = [expr];` (use getTargetTypeForAssignmentContext)
```

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 5]
```diff
@@ -1816,42 +1810,261 @@ public void checkTypeParameterNullnessForConditionalExpression(
...
     if (parent instanceof AssignmentTree || parent instanceof VariableTree) {
-      return getTreeType(parent, state.withPath(parentPath));
+      return getTargetTypeForAssignmentContext(parent, parentState, calledFromDataflow);
+    }
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:1978]:
```suggestion
    }
    // 2b. `return [expr];` => target is containing method's formal return type
```

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 6]
```diff
@@ -1816,42 +1810,261 @@ public void checkTypeParameterNullnessForConditionalExpression(
...
+          NullabilityUtil.findEnclosingMethodOrLambdaOrInitializer(parentPath);
+      if (enclosingMethodOrLambda != null
+          && enclosingMethodOrLambda.getLeaf() instanceof MethodTree enclosingMethod) {
+        Symbol.MethodSymbol methodSymbol = ASTHelpers.getSymbol(enclosingMethod);
+        if (methodSymbol != null) {
+          return new TargetTypeAndAssignmentKind(methodSymbol.getReturnType(), false);
+        }
+      }
+      return new TargetTypeAndAssignmentKind(null, false);
+    }
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:1990]:
```suggestion
    }
    // 2c. `foo(..., [expr]);` => target is called method's formal argument type
```

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 7]
```diff
@@ -1816,42 +1810,261 @@ public void checkTypeParameterNullnessForConditionalExpression(
...
+      }
+      Type methodType = ASTHelpers.getType(parentInvocation.getMethodSelect());
+      if (methodType == null) {
+        return new TargetTypeAndAssignmentKind(null, false);
+      }
+      return new TargetTypeAndAssignmentKind(
+          getFormalParameterTypeForArgument(
+              parentInvocation, methodType.asMethodType(), expressionTree),
+          false);
+    }
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:2006]:
```suggestion
    }
    // (Also 2c, but for constructors)
```

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 8]
```diff
@@ -1816,42 +1810,261 @@ public void checkTypeParameterNullnessForConditionalExpression(
...
+        Type parentCtorType =
+            TypeSubstitutionUtils.memberType(
+                state.getTypes(), parentClassType, parentCtorSymbol, config);
+        return new TargetTypeAndAssignmentKind(
+            getFormalParameterTypeForArgument(
+                parentConstructorCall, parentCtorType.asMethodType(), expressionTree),
+            false);
+      }
+    }
+    return new TargetTypeAndAssignmentKind(null, false);
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:2022]:
```suggestion
    // Fall back to an unknown target type
    return new TargetTypeAndAssignmentKind(null, false);
```

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 9]
```diff
@@ -1816,42 +1810,261 @@ public void checkTypeParameterNullnessForConditionalExpression(
...
+    Type condExprType =
+        inferConditionalExpressionType(
+            state,
+            tree,
+            state.getPath(),
+            targetTypeAndAssignmentKind.typeFromAssignmentContext(),
+            false);
+    if (!hasTargetType) {
+      return;
+    }
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java]: I assume there is some side-effect from `inferConditionalExpressionType` we want here, so that we don't do the check for `!hasTargetType` first. But, if so, I'd expect it to be documented in a comment. As it is, I am tempted to move the check _before_ computing `condExprType`.

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 10]
```diff
@@ -1816,42 +1810,261 @@ public void checkTypeParameterNullnessForConditionalExpression(
...
+
+  private record TargetTypeAndAssignmentKind(
+      @Nullable Type typeFromAssignmentContext, boolean assignedToLocal) {}
+
+  private TargetTypeAndAssignmentKind getTargetTypeForConditionalExpression(
+      ConditionalExpressionTree tree, VisitorState state, boolean calledFromDataflow) {
+    TreePath conditionalPath =
+        Objects.equals(state.getPath().getLeaf(), tree)
+            ? state.getPath()
+            : pathWithLeaf(state.getPath(), tree);
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java]: Why is state.getPath() potentially at two different places (and only those two places) here?

File: nullaway/src/test/java/com/uber/nullaway/jspecify/GenericDiamondTests.java [hunk 11]
```diff
@@ -44,6 +44,36 @@ void testPositive() {
...
+                Box(T value) {
+                  this.value = value;
+                }
+                T get() {
+                  return value;
+                }
+              }
+              void test() {
+                // BUG: Diagnostic contains: passing @Nullable parameter
+                var inferredFromInitializer = new Box<>(null);
```
Comment (lazaroclapp) [inline nullaway/src/test/java/com/uber/nullaway/jspecify/GenericDiamondTests.java:69]: Why does this return an error? Shouldn't it just infer `Box<@Nullable Object>`? (I think the test is explicitly asserting that it does not, but I am not sure I recall why those would be the semantics. Maybe worth a comment or JSpecify link?)

File: nullaway/src/test/java/com/uber/nullaway/jspecify/ConditionalExprTests.java [hunk 12]
```diff
@@ -0,0 +1,425 @@
...
+            final class Test {
+              String nonNullField = "hello";
+              @Nullable String nullableField;
+              static <T extends @Nullable Object> T id(T t) {
+                return t;
+              }
+              void test(boolean flag) {
+                nullableField = flag ? id(null) : id("fallback");
+                // BUG: Diagnostic contains: passing @Nullable parameter
+                nonNullField = flag ? id(null) : id("fallback");
```
Comment (lazaroclapp) [inline nullaway/src/test/java/com/uber/nullaway/jspecify/ConditionalExprTests.java:79]: Silly, but, for completeness, can we add the case where the null and non-null arms reversed too?

File: nullaway/src/test/java/com/uber/nullaway/jspecify/ConditionalExprTests.java [hunk 13]
```diff
@@ -0,0 +1,425 @@
...
+            final class Test {
+              String nonNullField = "hello";
+              @Nullable String nullableField;
+              static <T extends @Nullable Object> T id(T t) {
+                return t;
+              }
+              void test(boolean first, boolean second) {
+                nullableField = first ? id(null) : second ? id("value") : "fallback";
+                // BUG: Diagnostic contains: passing @Nullable parameter
+                nonNullField = first ? id(null) : second ? id("value") : "fallback";
```
Comment (lazaroclapp) [inline nullaway/src/test/java/com/uber/nullaway/jspecify/ConditionalExprTests.java:107]: Again, let's make sure it works with the second `id(...)` being the one that takes `null`

(I don't think we need this on every test, necessarily, but at least one plain and one nested case)

File: nullaway/src/test/java/com/uber/nullaway/jspecify/ConditionalExprTests.java [hunk 14]
```diff
@@ -0,0 +1,425 @@
...
+                Box(T value) {
+                  this.value = value;
+                }
+                T get() {
+                  return value;
+                }
+              }
+              void test(boolean flag) {
+                // BUG: Diagnostic contains: passing @Nullable parameter
+                var inferredFromInitializer = flag ? new Box<>(null) : new Box<>("fallback");
```
Comment (lazaroclapp) [inline nullaway/src/test/java/com/uber/nullaway/jspecify/ConditionalExprTests.java:280]: I mention it above, but again mildly surprised that `var inferredFromInitializer` isn't just inferred to be `Box<@Nullable String>` here. I suspect there is a reason for it in the JSpecify semantics, just think it might be worth a short comment (see `GenericDiamondTests.java`'s comment)

**Review by lazaroclapp — COMMENTED** [review]
Hey, sorry this one took me a while, but it is was indeed a fairly involved PR, so no chance to get to it during the week...

As it is, I still don't really have the entire algorithm / design swapped in memory (nor do I think I can in a feasible amount of time 😅 ), so a lot of my comments are local to what I see each individual method do and to the tests, but hopefully still useful!

Looks like we can handle a ton of complex inference cases now, though! Cool! 🎉

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java (hunk 1 above)
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java]: This is correct; will remove this one

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java (hunk 2 above)
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java]: Yes, that is better.  Will fix

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java (hunk 9 above)
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java]: No you're right, this code is just weird.  I moved the check as you suggested.

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java (hunk 10 above)
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java]: Sorry for this, this code should just call `pathWithLeaf` which already checks for equality of the leaf of the path; will fix.

File: nullaway/src/test/java/com/uber/nullaway/jspecify/GenericDiamondTests.java (hunk 11 above)
Comment (msridhar) [reply nullaway/src/test/java/com/uber/nullaway/jspecify/GenericDiamondTests.java:69]: You're right, there should be no warning there.  I opened a new issue for this, #1633.  I thought this was testing something useful before despite the limitation; looking more closely, I'm now not so sure.  But I want to leave this test in for now, to be sure it stays around.  So I'm just going to add a comment linking that new issue.

File: nullaway/src/test/java/com/uber/nullaway/jspecify/ConditionalExprTests.java (hunk 14 above)
Comment (msridhar) [reply nullaway/src/test/java/com/uber/nullaway/jspecify/ConditionalExprTests.java:280]: I've added a comment for this one as well

**Comment by msridhar** [comment]
Thanks for the detailed review @lazaroclapp !  I believe I've addressed all the feedback now, so this is ready for another look

**Review by lazaroclapp — APPROVED** [review]
